### PR TITLE
Dt 603 validation for court hearings

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Initial PR to add validation to the scheduling of court hearings.

### CODE REVIEW

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 1]
```diff
@@ -45,33 +50,65 @@ public CourtHearingsService(final OffenderBookingRepository offenderBookingRepos
...
+
+    private void checkPrisonLocationSameAsOffenderBooking(final String prisonLocation, final OffenderBooking booking) {
+        var agency = agencyLocationRepository.findById(prisonLocation).orElseThrow(EntityNotFoundException.withId(prisonLocation));
+
+        if (!booking.getLocation().equals(agency)) {
+            throw new IllegalArgumentException("Prison location does not match the bookings location.");
+        }
+    }
+
+    // TODO check if court is active.
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: I know this is here.

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 2]
```diff
@@ -45,33 +50,65 @@ public CourtHearingsService(final OffenderBookingRepository offenderBookingRepos
...
+    }
+
+    private OffenderCourtCase getActiveCourtCaseFor(final Long caseId, final OffenderBooking offenderBooking) {
+        final var courtCase = offenderBooking.getCourtCaseBy(caseId).orElseThrow(EntityNotFoundException.withId(caseId));
+
+        if (courtCase.isActive()) {
+            return courtCase;
+        }
+
+        throw new IllegalArgumentException(format("Court case with id [%d] is not active.", caseId));
```
Comment (andrewrlee) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: these could potentially all be replaced by `checkState`s?

File: src/test/java/net/syscon/elite/service/CourtHearingsServiceTest.java [hunk 3]
```diff
@@ -108,22 +128,130 @@ void setup() {
...
+        givenProvidedCourtLocationIsNotACourt();
+
+        assertThatThrownBy(() -> courtHearingsService.scheduleHearing(OFFENDER_BOOKING_ID, COURT_HEARING))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Supplied court location wih id COURT is not a valid court location.");
+    }
+
+    private void givenProvidedCourtLocationIsNotACourt() {
+        when(offenderBooking.isActive()).thenReturn(true);
+        when(offenderBooking.getLocation()).thenReturn(fromPrison);
```
Comment (andrewrlee) [inline src/test/java/net/syscon/elite/service/CourtHearingsServiceTest.java]: mocking out a load of stuff - isn't it easier to instantiate these objects using their builders and populate so they are in the correct state?

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java (hunk 2 above)
Comment (andrewrlee) [reply src/main/java/net/syscon/elite/service/CourtHearingsService.java]: apache commons `Validate` is used quite a bit too.

File: src/test/java/net/syscon/elite/service/CourtHearingsServiceTest.java (hunk 3 above)
Comment (matt-crowson-moj) [reply src/test/java/net/syscon/elite/service/CourtHearingsServiceTest.java]: Yes I know.  Gonna tidy this up. I realise it is ugly.

Comment (andrewrlee) [reply src/test/java/net/syscon/elite/service/CourtHearingsServiceTest.java]: Theres a nice pattern they use in community-api. 
Create a method that produces an exemplar  
Then in test call method, then `toBuilder` and configure object with bits salient for test:

e.g: https://github.com/ministryofjustice/community-api/blob/master/src/test/java/uk/gov/justice/digital/delius/service/StaffServiceTest.java#L147

Comment (matt-crowson-moj) [reply src/test/java/net/syscon/elite/service/CourtHearingsServiceTest.java]: OK well I had already made changes.  I think will suffice for now (hint hint).

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/java/net/syscon/elite/service/CourtHearingsService.java]: Done.  Updated to use Preconditions.

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 4]
```diff
@@ -45,31 +50,59 @@ public CourtHearingsService(final OffenderBookingRepository offenderBookingRepos
...
+    private OffenderCourtCase getActiveCourtCaseFor(final Long caseId, final OffenderBooking offenderBooking) {
+        final var courtCase = offenderBooking.getCourtCaseBy(caseId).orElseThrow(EntityNotFoundException.withId(caseId));
+
+        checkArgument(courtCase.isActive(), "Court case with id %s is not active.", caseId);
+
+        return courtCase;
+    }
+
+    private void checkPrisonLocationSameAsOffenderBooking(final String prisonLocation, final OffenderBooking booking) {
+        var agency = agencyLocationRepository.findById(prisonLocation).orElseThrow(EntityNotFoundException.withId(prisonLocation));
```
Comment (andrewrlee) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: meh.. missing final

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 5]
```diff
@@ -45,31 +50,59 @@ public CourtHearingsService(final OffenderBookingRepository offenderBookingRepos
...
+
+        checkArgument(booking.getLocation().equals(agency),"Prison location does not match the bookings location.");
+    }
+
+    // TODO check if court is active.
+    private AgencyLocation getCourtFor(final String courtLocation) {
+        var agency = agencyLocationRepository.findById(courtLocation).orElseThrow(EntityNotFoundException.withId(courtLocation));
```
Comment (andrewrlee) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: missing final

**Pull request merged**